core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/stopstartds/StopStartDownscaleActions.java

```diff
@@ -86,15 +87,34 @@ protected void prepareExecution(StopStartDownscaleTriggerEvent payload, Map<Obje
             }
 
             @Override
-            protected void doExecute(
-                    StopStartDownscaleContext context, StopStartDownscaleTriggerEvent payload, Map<Object, Object> variables) throws Exception {
-                stopStartDownscaleFlowService.clusterDownscaleStarted(context.getStack().getId(), payload.getHostGroupName(), payload.getHostIds());
-                sendEvent(context);
+            protected void doExecute(StopStartDownscaleContext context, StopStartDownscaleTriggerEvent payload, Map<Object, Object> variables) throws Exception {
+                Stack stack = context.getStack();
+
+                List<InstanceMetaData> allInstanceMetadataInHostGroup = stack.getNotDeletedAndNotZombieInstanceMetaDataList().stream()
+                        .filter(i -> i.getInstanceGroupName().equals(context.getHostGroupName()))
+                        .collect(Collectors.toList());
+
+                List<CloudInstance> allInstancesInHostGroup = instanceMetaDataToCloudInstanceConverter.convert(allInstanceMetadataInHostGroup,
+                        stack.getEnvironmentCrn(), stack.getStackAuthentication());
+                StopStartDownscaleGetRunningInstancesRequest request = new StopStartDownscaleGetRunningInstancesRequest(context.getCloudContext(),
+                        context.getCloudCredential(), context.getCloudStack(), context.getHostGroupName(), allInstancesInHostGroup, payload.getHostIds());
+                sendEvent(context, request);
             }
+        };
+    }
+
+    // TODO CB-15132: Potential pre-flight checks. YARN / appropriate service masters available. CM master available.
+    @Bean(name = "STOPSTART_DOWNSCALE_HOSTS_DECOMMISSION_STATE")
+    public Action<?, ?> decommissionViaCmAction() {
+        return new AbstractStopStartDownscaleActions<>(StopStartDownscaleGetRunningInstancesResult.class) {
 
             @Override
-            protected Selectable createRequest(StopStartDownscaleContext context) {
-                return new StopStartDownscaleDecommissionViaCMRequest(context.getStack().getId(), context.getHostGroupName(), context.getHostIdsToRemove());
+            protected void doExecute(
+                    StopStartDownscaleContext context, StopStartDownscaleGetRunningInstancesResult payload, Map<Object, Object> variables) throws Exception {
+                stopStartDownscaleFlowService.clusterDownscaleStarted(context.getStack().getId(), payload.getHostGroupName(), payload.getHostIds());
```

Comment:
Messages here will need to be changed appropriately to include information about the nodes which were selected for recovery, if any (only to be logged if such nodes were found). Also the language on the messages - related to 'Starting downscale' etc need review to make sure they still make sense.